builtin/providers/openstack/resource_openstack_compute_instance_v2.go

```diff
@@ -0,0 +1,1011 @@
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+				ForceNew: false,
+			},
+			"key_pair": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"block_device": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"uuid": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"source_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"volume_size": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"destination_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"boot_index": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"volume": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+						"volume_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"device": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+				Set: resourceComputeVolumeAttachmentHash,
+			},
+		},
+	}
+}
+
+func resourceComputeInstanceV2Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	computeClient, err := config.computeV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack compute client: %s", err)
+	}
+
+	var createOpts servers.CreateOptsBuilder
+
+	imageId, err := getImageID(computeClient, d)
+	if err != nil {
+		return err
+	}
+
+	flavorId, err := getFlavorID(computeClient, d)
+	if err != nil {
+		return err
+	}
+
+	createOpts = &servers.CreateOpts{
+		Name:             d.Get("name").(string),
+		ImageRef:         imageId,
+		FlavorRef:        flavorId,
+		SecurityGroups:   resourceInstanceSecGroupsV2(d),
+		AvailabilityZone: d.Get("availability_zone").(string),
+		Networks:         resourceInstanceNetworksV2(d),
+		Metadata:         resourceInstanceMetadataV2(d),
+		ConfigDrive:      d.Get("config_drive").(bool),
+		AdminPass:        d.Get("admin_pass").(string),
+	}
+
+	if keyName, ok := d.Get("key_pair").(string); ok && keyName != "" {
+		createOpts = &keypairs.CreateOptsExt{
+			createOpts,
+			keyName,
+		}
+	}
+
+	if blockDeviceRaw, ok := d.Get("block_device").(map[string]interface{}); ok && blockDeviceRaw != nil {
+		blockDevice := resourceInstanceBlockDeviceV2(d, blockDeviceRaw)
+		createOpts = &bootfromvolume.CreateOptsExt{
+			createOpts,
+			blockDevice,
+		}
+	}
+
+	log.Printf("[INFO] Requesting instance creation")
```

Comment:
One thing that has really helped us with debugging over in the other providers is doing a debug dump of the API request just before we make it. Here's [an example of it in the AWS provider](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_instance.go#L391).

You may want to consider doing simpler `fmt.Printf("%#v")` style outputs of your `*Opts` vars in your resources. :+1: